text/0605-eks-rewrite.md

```diff
@@ -0,0 +1,437 @@
+be created by default. It will only be deployed when `kubectlProviderOptions`
+property is used.**
+
+```
+new eks.Cluster(this, 'hello-eks', {
+  version: eks.KubernetesVersion.V1_31,
+  # Using this property will create `Kubectl Handler` as custom resource handler
+  kubectlProviderOptions: {
+    kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  }
+});
+```
+
+### VPC Support
+
+You can specify the VPC of the cluster using the vpc and vpcSubnets properties:
+
+```
+declare const vpc: ec2.Vpc;
+new eks.Cluster(this, 'HelloEKS', {
+  version: eks.KubernetesVersion.V1_31,
+  vpc,
+  vpcSubnets: [{ subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS }],
+});
+```
+
+If you do not specify a VPC, one will be created on your behalf, which you can then access via cluster.vpc.
+The cluster VPC will be associated to any EKS managed capacity (i.e Managed Node Groups and Fargate Profiles).
+
+The cluster can be placed inside an isolated VPC. The cluster’s VPC subnets must have a VPC interface endpoint
+for any AWS services that your Pods need access to. See https://docs.aws.amazon.com/eks/latest/userguide/private-clusters.html
+
+```
+const vpc = new ec2.Vpc(this, 'vpc', {
+    subnetConfiguration: [
+        {
+            cidrMask: 24,
+            name: 'Private',
+            subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
+        },
+    ]
+});
+
+const cluster = new eks.Cluster(this, 'Mycluster', {
+    version: eks.KubernetesVersion.V1_31,
+    vpc,
+    vpcSubnets: [{ subnetType: ec2.SubnetType.PRIVATE_ISOLATED }]
+});
+```
+
+### Kubectl Support
+
+You can choose to have CDK create a `Kubectl Handler` - a Python Lambda Function to
+apply k8s manifests using `kubectl apply`. This handler will not be created by default.
+
+To create a `Kubectl Handler`, use `kubectlProviderOptions` when creating the cluster.
+`kubectlLayer` is the only required property in `kubectlProviderOptions`.
+
+```
+new eks.Cluster(this, 'hello-eks', {
+  version: eks.KubernetesVersion.V1_31,
+  # Using this property will create `Kubectl Handler` as custom resource handler
+  kubectlProviderOptions: {
+    kubectlLayer: new KubectlV31Layer(this, 'kubectl'),
+  }
+});
+
+`Kubectl Handler` created along with the cluster will be granted admin permissions to the cluster.
+```
+
+## Permissions and Security
+
+Amazon EKS supports three modes of authentication: CONFIG_MAP, API_AND_CONFIG_MAP, and API.
+`ConfigMap` authentication mode has been deprecated by EKS and the recommended mode is API.
+The new EKS L2 will go a step further and only support API authentication mode.
+All grant functions in EKS will use Access Entry to grant
+permissions to an IAM role/user.
+
+As a result, `AwsAuth` construct in the previous EKS module will not be provided in the new module.
+`grant()` functions are introduced to replace the awsAuth. It’s implemented using
+Access Entry.
+
+Grant Admin Access to an IAM role
+
+```
+cluster.grantAdmin('adminAccess', roleArn, eks.AccessScopeType.CLUSTER);
```

Comment:
Can we make the second argument here accept an [IGrantable](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/lib/principals.ts#L14) and not a string?

Reply:
Does this function return something? an `AccessEntry` maybe?

Reply:
This API is basically the current `grantAccess()` with hardcode values: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_eks.Cluster.html#grantwbraccessid-principal-accesspolicies

> Can we make the second argument here accept an [IGrantable](https://github.com/aws/aws-cdk/blob/main/packages/aws-cdk-lib/aws-iam/lib/principals.ts#L14) and not a string?

Yes I agree.

> Does this function return something? an `AccessEntry` maybe?

It's the same as current API which is `void`.

> Is the access scope argument mandatory? what would be the API for granting admin on a specific namespace?


Actually I think we should remove the third argument to make `grantAdmin()` easier to use. So something like:
```
cluster.grantAdmin('adminAccess', role);
```
It will add `AmazonEKSClusterAdminPolicy` to the role. There's also AmazonEKSAdminPolicy which can be applied to a namespace. Have 2 `grantAdmin` is a bit confusing. Customers can use the general grant method for granting admin on a specific namespace. We can add more`grantXXX()` later.